From c15c02b69df9a998d644d0c91852a4baf260892a Mon Sep 17 00:00:00 2001
From: Athanasios Oikonomou <athoik@gmail.com>
Date: Sun, 11 May 2014 12:13:46 +0300
Subject: [PATCH] Handle DTS Passthrough and Downmix

When downmix is enabled we remove on runtime the dts caps to give priority on dtsdownmix.

diff --git a/src/gstdvbaudiosink.c b/src/gstdvbaudiosink.c
index 46df63a..118d1fd 100644
--- a/src/gstdvbaudiosink.c
+++ b/src/gstdvbaudiosink.c
@@ -62,6 +62,7 @@
 #endif
 #include <unistd.h>
 #include <stdint.h>
+#include <stdio.h>
 #include <string.h>
 #include <sys/ioctl.h>
 #include <sys/socket.h>
@@ -297,6 +298,19 @@ gst_dvbaudiosink_base_init (gpointer klass)
 	gst_element_class_set_details (element_class, &element_details);
 }
 
+static gboolean get_downmix_setting()
+{
+	FILE *f;
+	char buffer[32] = {0};
+	f = fopen("/proc/stb/audio/ac3", "r");
+	if (f)
+	{
+		fread(buffer, sizeof(buffer), 1, f);
+		fclose(f);
+	}
+	return !strncmp(buffer, "downmix", 7);
+}
+
 static GstCaps *
 gst_dvbaudiosink_get_caps (GstBaseSink *basesink)
 {
@@ -325,7 +339,12 @@ gst_dvbaudiosink_get_caps (GstBaseSink *basesink)
 		gst_caps_remove_structure(caps, 12); // remove x-private-eac3
 	}
 	else
-		caps = gst_static_caps_get(&hwtemplate->static_caps);
+		caps = gst_caps_copy(&hwtemplate->static_caps.caps);
+
+	if (hwtemplate == &sink_factory_broadcom_dts && get_downmix_setting()) {
+		gst_caps_remove_structure(caps, 9); // remove x-private1-dts
+		gst_caps_remove_structure(caps, 8); // remove x-dts
+	}
 
 //	strcaps = gst_caps_to_string(caps);
 //	GST_INFO_OBJECT (self, "dynamic caps for model %d '%s'", hwtype, gst_caps_to_string(caps));
-- 
1.7.10.4

